process-migration-service/src/main/java/org/kie/processmigration/service/impl/MigrationServiceImpl.java

```diff
@@ -0,0 +1,200 @@
+        return migrate(migration, plan, credentials);
+    }
+
+    private Migration migrate(Migration migration, Plan plan, Credentials credentials) {
+        try {
+            migration = em.find(Migration.class, migration.getId());
+            em.persist(migration.start());
+            if (credentials == null) {
+                throw new CredentialNotFoundException();
+            }
+            AtomicBoolean hasErrors = new AtomicBoolean(false);
+            //each instance id will spawn its own request to KIE server for migration.
+            List<Long> instancesIdList = migration.getDefinition().getProcessInstanceIds();
+            for (Long instanceId : instancesIdList) {
+                MigrationReportInstance reportInstance = kieService
+                                                                   .createProcessAdminServicesClient(CredentialsProviderFactory.getProvider(credentials))
+                                                                   .migrateProcessInstance(
+                                                                                           plan.getSourceContainerId(),
+                                                                                           instanceId,
+                                                                                           plan.getTargetContainerId(),
+                                                                                           plan.getTargetProcessId(),
+                                                                                           plan.getMappings());
+                if (!hasErrors.get() && !reportInstance.isSuccessful()) {
+                    hasErrors.set(Boolean.TRUE);
+                }
+                em.persist(new MigrationReport(migration.getId(), reportInstance));
+            }
+            migration.complete(hasErrors.get());
+        } catch (Exception e) {
+            logger.error("Migration failed", e);
+            migration.fail(e);
+        } finally {
+            credentialsService.delete(migration.getId());
+            em.persist(migration);
+            if (ExecutionType.ASYNC.equals(migration.getDefinition().getExecution().getType()) &&
+                migration.getDefinition().getExecution().getCallbackUrl() != null) {
+                doCallback(migration);
+            }
+        }
+        return migration;
+    }
+
+    private void doCallback(Migration migration) {
+        URI callbackURI = null;
+        try {
+            callbackURI = migration.getDefinition().getExecution().getCallbackUrl();
+            Response response = ClientBuilder.newClient()
+                                             .target(callbackURI)
+                                             .request(MediaType.APPLICATION_JSON)
+                                             .buildPost(Entity.json(migration))
+                                             .invoke();
```

Comment:
missing authentication support ... so all callbacks must be open endpoints

Reply:
Do you suggest to add support to basic auth?

Reply:
The callbackUri can include a (unique?) token e.g. http://somehost.example.com/callback?token=1sdf2342fsf